### PR TITLE
[Model] Full ABINet Framework

### PR DESCRIPTION
**Pull request by gaotongxiao**
Implementation of [Read Like Humans: Autonomous, Bidirectional and Iterative Language Modeling for Scene Text Recognition](https://arxiv.org/abs/2103.06495)

### CODE REVIEW

File: mmocr/models/textrecog/decoders/abinet_language_decoder.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                 pad_idx=0,
+                 init_cfg=None,
+                 **kwargs):
+        super().__init__(init_cfg=init_cfg)
+        self.detach_tokens = detach_tokens
+
+        self.d_model = d_model
+        self.max_seq_len = max_seq_len
+
+        self.proj = nn.Linear(num_chars, d_model, False)
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/decoders/abinet_language_decoder.py:57]: self.project or self.fc

File: mmocr/models/textrecog/decoders/abinet_language_decoder.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        out = (logit.argmax(dim=-1) == self.pad_idx)
+        abn = out.any(dim)
+        # Get the first index of end token
+        out = ((out.cumsum(dim) == 1) & out).max(dim)[1]
+        out = out + 1
+        out = torch.where(abn, out, out.new_tensor(logit.shape[1]))
+        return out
+
+    @staticmethod
+    def _get_location_mask(sz, device=None):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/decoders/abinet_language_decoder.py:151]: docstring

File: mmocr/models/textrecog/decoders/abinet_language_decoder.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        mask = torch.eye(sz, device=device)
+        mask = mask.float().masked_fill(mask == 1, float('-inf'))
+        return mask
+
+    @staticmethod
+    def _get_padding_mask(length, max_length):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/decoders/abinet_language_decoder.py:157]: docstring

File: mmocr/models/textrecog/losses/mix_loss.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        self.dec_weight = dec_weight
+        self.fusion_weight = fusion_weight
+        self.num_classes = num_classes
+
+    def _flatten(self, logits, target_lens):
+        flatten_logits = torch.cat(
+            [s[:target_lens[i]] for i, s in enumerate((logits))])
+        return flatten_logits
+
+    def ce(self, logits, targets):
```
Comment (Harold-lkk) [inline mmocr/models/textrecog/losses/mix_loss.py:46]: ce_loss

File: mmocr/models/builder.py [hunk 5]
```diff
@@ -20,6 +20,7 @@
 ROI_EXTRACTORS = BACKBONES
 HEADS = BACKBONES
 NECKS = BACKBONES
+FUSERS = BACKBONES
```
[the remark this answers is not in the document]
Comment (gaotongxiao) [reply mmocr/models/builder.py:23]: Naming it with "FUSERS" is consistent with the practice of "ENCODERS", "DECODERS" and "CONVERTERS", etc. It implies the module itself is more like an executor. "FUSIONS" is just a noun describing an objective process.

[the remark this answers is not in the document]
Comment (Harold-lkk) [reply mmocr/models/builder.py:23]: reasonable

**Review by Harold-lkk — APPROVED** [review]

**Pull request merged**